server/src/main/java/org/elasticsearch/index/codec/vectors/DefaultIVFVectorsWriter.java

```diff
@@ -415,197 +264,40 @@ protected int calculateAndWriteCentroids(
-                int sz = neighborsToCheck.size();
-                int best = neighborsToCheck.consumeNodesAndScoresMin(ordScoreIterator.ords, ordScoreIterator.scores);
-                // Set the size to the number of neighbors we actually found
-                ordScoreIterator.setSize(sz);
-                bestScore = ordScoreIterator.getScore(best);
-                bestCentroid = ordScoreIterator.getOrd(best);
             }
-            clusters[bestCentroid].add(docID);
-            if (soarClusterCheckCount > 0) {
-                assignCentroidSOAR(
-                    ordScoreIterator,
-                    docID,
-                    bestCentroid,
-                    scorer.centroid(bestCentroid),
-                    bestScore,
-                    scratch,
-                    scorer,
-                    vector,
-                    clusters
-                );
-            }
-            neighborsToCheck.clear();
-        }
-    }
-
-    static void assignCentroidSOAR(
-        OrdScoreIterator centroidsToCheck,
-        int vecOrd,
-        int bestCentroidId,
-        float[] bestCentroid,
-        float bestScore,
-        float[] scratch,
-        CentroidAssignmentScorer scorer,
-        float[] vector,
-        IntArrayList[] clusters
-    ) throws IOException {
-        ESVectorUtil.subtract(vector, bestCentroid, scratch);
-        int bestSecondaryCentroid = -1;
-        float minDist = Float.MAX_VALUE;
-        for (int i = 0; i < centroidsToCheck.size(); i++) {
-            float score = centroidsToCheck.getScore(i);
-            int centroidOrdinal = centroidsToCheck.getOrd(i);
-            if (centroidOrdinal == bestCentroidId) {
-                continue;
-            }
-            float proj = ESVectorUtil.soarResidual(vector, scorer.centroid(centroidOrdinal), scratch);
-            score += SOAR_LAMBDA * proj * proj / bestScore;
-            if (score < minDist) {
-                bestSecondaryCentroid = centroidOrdinal;
-                minDist = score;
-            }
-        }
-        if (bestSecondaryCentroid != -1) {
-            clusters[bestSecondaryCentroid].add(vecOrd);
-        }
-    }
-
-    static class OrdScoreIterator {
-        private final int[] ords;
-        private final float[] scores;
-        private int idx = 0;
 
-        OrdScoreIterator(int size) {
-            this.ords = new int[size];
-            this.scores = new float[size];
-        }
-
-        int setSize(int size) {
-            if (size > ords.length) {
-                throw new IllegalArgumentException("size must be <= " + ords.length);
+            for (int j = 0; j < soarAssignments.length; j++) {
+                if (soarAssignments[j] == c) {
+                    cluster.add(j);
+                }
             }
-            this.idx = size;
-            return size;
-        }
 
-        int getOrd(int idx) {
-            return ords[idx];
+            assignmentsByCluster[c] = cluster;
```

Comment:
Can we call `cluster.trimToSize()` before so we remove any extra empty slot? I think it can save some heap here.